### DISCUSSION TITLE
can we implement one tilesetdef in another?

### DISCUSSION THREAD

**Discussion by BuzzLEquerre** [flareteam/flare-game#967] [General]
Hello, I have a project to merge my tilesetdef of the caves and that of the plains. I would like to know if I have to recode everything by adjusting the tile numbers or if I can via a method readjust all of my tiles
[tileset_cave.txt](https://github.com/user-attachments/files/16753285/tileset_cave.txt)
[tileset_grassland.txt](https://github.com/user-attachments/files/16753286/tileset_grassland.txt)

**Comment by dorkster** [comment]
Unfortunately, there's no quick method here. You'll have to adjust the tile IDs.

If you open the .tmx file in a text editor, you should be able to see the `firstgid` property for each tileset. This will tell you where to start the numbering for a particular tileset.

**Reply by BuzzLEquerre** [reply]
Hello, I have readjusted my cave and room tiles, but I have a shift of 43 tiles in the cave tileset and I don't see how and where to fix it.
instead of having 1963 for the first tile of the room, I have 2006 :
![Capture d'écran 2024-08-27 101049](https://github.com/user-attachments/assets/146bf739-e186-4748-9db6-76ea3468084f)

If you have any idea why I have such a big lag, I would like your help.
[tileset_cave.txt](https://github.com/user-attachments/files/16758863/tileset_cave.txt)
[tileset_grassland.txt](https://github.com/user-attachments/files/16758866/tileset_grassland.txt)

**Reply by BuzzLEquerre** [reply]
screenshots :
![Capture d'écran 2024-08-27 102110](https://github.com/user-attachments/assets/01aecf11-1434-4628-b3a4-1854e7014a9a)
![Capture d'écran 2024-08-27 102143](https://github.com/user-attachments/assets/6d309e80-d93d-4d19-bede-52362d0939fa)

**Reply by dorkster** [reply]
Can you share the map file in both Tiled tmx and Flare txt formats?

**Reply by BuzzLEquerre** [reply]
Github dont support the tmx file type i shre you a screenshot of the Tiled map editor look 
[the_lost_island.txt](https://github.com/user-attachments/files/16763917/the_lost_island.txt)
![Capture d'écran 2024-08-27 164454](https://github.com/user-attachments/assets/da98e298-2f2b-44d3-b113-1dd8650b6b69)

I also placed my tilesets in the collision layer so you can easily observe my tile numbers

**Reply by dorkster** [reply]
You can either rename the tmx file extension to txt (it's just an XML file, so it'll still readable), or you can put it in a zip file. Github should support both.

**Reply by BuzzLEquerre** [reply]
hello, I can't change the file format but that's okay, I'm going to start adding my tilesets again,
and I have an off-topic question:

![Capture d'écran 2024-08-28 095119](https://github.com/user-attachments/assets/8e978caf-6f75-4f09-9486-6f0f9900f4d0)
![Capture d'écran 2024-08-28 095138](https://github.com/user-attachments/assets/b72af327-1f6a-48f3-944e-68504b381452)
Yesterday evening I started modifying the interface and I would like to know if I can put the same player animation in the game launch screen in the inventory, so that the player sees his hero

**Reply by dorkster** [reply]
Not possible at the moment. Having the character model in the inventory is an interesting idea, but it's not something we ever felt was necessary to add:

1. The player model is already visible in the center of the screen, and the inventory typically doesn't overlap it.
2. For flare-game in particular, the inventory menu is already fairly cramped. I think having the player model there would be too much. This might not apply to games that use are larger menu for the inventory.

One more thing on the tilesets: You can try using the [flare-tiled-tools](https://github.com/flareteam/flare-tiled-tools) extension for Tiled to export a tileset. You won't be able to use it directly, because the tilesets are packed for flare-game. But you can use it as a reference for the tile IDs and compare it to your edited tileset definition.

**Reply by BuzzLEquerre** [reply]
Yes, but logically it shouldn't be very difficult to add ? Since in the menu we see our player, is it enough to add a function or a command in the game engine to display the player in the game inventory ?

**Reply by dorkster** [reply]
Not too difficult at all :)
https://github.com/flareteam/flare-engine/commit/0941bafc1a2346302d43ddeb50e0ff271e0ae63b

https://github.com/user-attachments/assets/71c3f63d-5d56-47c3-ac8c-91f8f74bf1b0

**Reply by BuzzLEquerre** [reply]
It's perfect !!
I just finished the layout of my inventory design,
thank you so mush !
![inventory](https://github.com/user-attachments/assets/4aee0955-25d7-42c1-9c36-7a00b1e22a77)
(and I can change the layout of the inventory image very easily, I use photoshop)

**Reply by BuzzLEquerre** [reply]
Hello, I added the lines of code that you did to display the player in the inventory but when I execute "mingw32-make" it does not work :


> 
> MSI Dragon@PC-Romain CLANG64 ~/flare-engine
> $ mingw32-make
> [  1%] Building CXX object CMakeFiles/flare.dir/src/Avatar.cpp.obj
> [  2%] Building CXX object CMakeFiles/flare.dir/src/CampaignManager.cpp.obj
> [  3%] Building CXX object CMakeFiles/flare.dir/src/EventManager.cpp.obj
> [  4%] Building CXX object CMakeFiles/flare.dir/src/GameSlotPreview.cpp.obj
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp: In member function 'void GameSlotPreview::addRenders(std::vector<Renderable>&)':
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:181:23: warning: declaration of 'i' shadows a previous local [-Wshadow]
>   181 |         for (unsigned i = 0; i < layer_def[dir].size(); ++i) {
>       |                       ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:175:23: note: shadowed declaration is here
>   175 |         for (unsigned i = 0; i < layer_def[stats->direction].size(); ++i) {
>       |                       ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:182:26: warning: declaration of 'index' shadows a previous local [-Wshadow]
>   182 |                 unsigned index = layer_def[dir][i];
>       |                          ^~~~~
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:176:26: note: shadowed declaration is here
>   176 |                 unsigned index = layer_def[stats->direction][i];
>       |                          ^~~~~
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:184:36: warning: declaration of 'ren' shadows a previous local [-Wshadow]
>   184 |                         Renderable ren = anims[index]->getCurrentFrame(dir);
>       |                                    ^~~
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:178:36: note: shadowed declaration is here
>   178 |                         Renderable ren = anims[index]->getCurrentFrame(stats->direction);
>       |                                    ^~~
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:178:36: warning: variable 'ren' set but not used [-Wunused-but-set-variable]
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:191:35: error: qualified-id in declaration before '(' token
>   191 | void GameSlotPreview::setStatBlock(StatBlock *_stats) {
>       |                                   ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:198:29: error: qualified-id in declaration before '(' token
>   198 | void GameSlotPreview::setPos(Point _pos) {
>       |                             ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:202:35: error: qualified-id in declaration before '(' token
>   202 | void GameSlotPreview::setDirection(unsigned char dir) {
>       |                                   ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:207:29: error: qualified-id in declaration before '(' token
>   207 | void GameSlotPreview::render() {
>       |                             ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:221:42: error: qualified-id in declaration before '(' token
>   221 | void GameSlotPreview::loadDefaultGraphics() {
>       |                                          ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:245:48: error: qualified-id in declaration before '(' token
>   245 | void GameSlotPreview::loadGraphicsFromInventory(MenuInventory* menu_inv) {
>       |                                                ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:279:36: error: expected ';' before '
> ' token
>   279 | GameSlotPreview::~GameSlotPreview() {
>       |                                    ^~
>       |                                    ;
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:290:2: error: expected '}' at end of input
>   290 | }
>       |  ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:175:75: note: to match this '{'
>   175 |         for (unsigned i = 0; i < layer_def[stats->direction].size(); ++i) {
>       |                                                                           ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:290:2: error: expected '}' at end of input
>   290 | }
>       |  ^
> C:\msys64\home\MSI Dragon\flare-engine\src\GameSlotPreview.cpp:171:62: note: to match this '{'
>   171 | void GameSlotPreview::addRenders(std::vector<Renderable> &r) {
>       |                                                              ^
> mingw32-make[2]: *** [CMakeFiles\flare.dir\build.make:406: CMakeFiles/flare.dir/src/GameSlotPreview.cpp.obj] Error 1
> mingw32-make[1]: *** [CMakeFiles\Makefile2:82: CMakeFiles/flare.dir/all] Error 2
> mingw32-make: *** [Makefile:135: all] Error 2
> 

The problem seems to come from GameSlotPreview, but I don't understand my mistake : 



/*
Copyright © 2011-2012 Clint Bellanger
Copyright © 2012 Igor Paliychuk
Copyright © 2012 Stefan Beller
Copyright © 2013 Henrik Andersson
Copyright © 2012-2016 Justin Jacobs

This file is part of FLARE.

FLARE is free software: you can redistribute it and/or modify it under the terms
of the GNU General Public License as published by the Free Software Foundation,
either version 3 of the License, or (at your option) any later version.

FLARE is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
FLARE.  If not, see http://www.gnu.org/licenses/
*/

/**
 * class GameSlotPreview
 *
 * Contains logic and rendering routines for the previews in GameStateLoad.
 */

#include "AnimationSet.h"
#include "CommonIncludes.h"
#include "GameSlotPreview.h"
#include "ItemManager.h"
#include "FileParser.h"
#include "MenuInventory.h"
#include "ModManager.h"
#include "RenderDevice.h"
#include "SharedResources.h"
#include "SharedGameResources.h"
#include "StatBlock.h"
#include "Utils.h"
#include "UtilsFileSystem.h"
#include "UtilsParsing.h"

GameSlotPreview::GameSlotPreview()
	: stats(NULL)
	, static_direction(6)
	, direction(&static_direction)
{
	// load the hero's animations from hero definition file
	anim->increaseCount("animations/hero.txt");
	animationSet = anim->getAnimationSet("animations/hero.txt");
	activeAnimation = animationSet->getAnimation("");

	loadLayerDefinitions();
}

/**
 * Load avatar sprite layer definitions into vector.
 */
void GameSlotPreview::loadLayerDefinitions() {
	// load layer definitions
	layer_def = std::vector<std::vector<unsigned> >(8, std::vector<unsigned>());
	layer_reference_order = std::vector<std::string>();

	// NOTE: This is documented in Avatar.cpp
	FileParser infile;
	if (infile.open("engine/hero_layers.txt", FileParser::MOD_FILE, FileParser::ERROR_NORMAL)) {
		while(infile.next()) {
			if (infile.key == "layer") {
				unsigned dir = Parse::toDirection(Parse::popFirstString(infile.val));
				if (dir>7) {
					infile.error("GameSlotPreview: Hero layer direction must be in range [0,7]");
					Utils::logErrorDialog("GameSlotPreview: Hero layer direction must be in range [0,7]");
					mods->resetModConfig();
					Utils::Exit(1);
				}
				std::string layer = Parse::popFirstString(infile.val);
				while (layer != "") {
					// check if already in layer_reference:
					unsigned ref_pos;
					for (ref_pos = 0; ref_pos < layer_reference_order.size(); ++ref_pos)
						if (layer == layer_reference_order[ref_pos])
							break;
					if (ref_pos == layer_reference_order.size())
						layer_reference_order.push_back(layer);
					layer_def[dir].push_back(ref_pos);

					layer = Parse::popFirstString(infile.val);
				}
			}
			else {
				infile.error("GameSlotPreview: '%s' is not a valid key.", infile.key.c_str());
			}
		}
		infile.close();
	}

	// There are the positions of the items relative to layer_reference_order
	// so if layer_reference_order=main,body,head,off
	// and we got a layer=3,off,body,head,main
	// then the layer_def[3] looks like (3,1,2,0)
}

void GameSlotPreview::loadGraphics(std::vector<std::string> _img_gfx) {
	if (!stats)
		return;

	loadLayerDefinitions();

	for (unsigned int i=0; i<animsets.size(); i++) {
		if (animsets[i])
			anim->decreaseCount(animsets[i]->getName());
		delete anims[i];
	}
	animsets.clear();
	anims.clear();

	for (unsigned int i=0; i<_img_gfx.size(); i++) {
		if (_img_gfx[i] != "") {
			std::string name = "animations/avatar/"+stats->gfx_base+"/"+_img_gfx[i] +".txt";
			anim->increaseCount(name);
			animsets.push_back(anim->getAnimationSet(name));
			animsets.back()->setParent(animationSet);
			anims.push_back(animsets.back()->getAnimation(activeAnimation->getName()));
			setAnimation("stance");
			if(!anims.back()->syncTo(activeAnimation)) {
				Utils::logError("GameSlotPreview: Error syncing animation in '%s' to 'animations/hero.txt'.", animsets.back()->getName().c_str());
			}
		}
		else {
			animsets.push_back(NULL);
			anims.push_back(NULL);
		}
	}
	anim->cleanUp();

	setAnimation("stance");
}

void GameSlotPreview::logic() {
	// handle animation
	activeAnimation->advanceFrame();
	for (unsigned i=0; i < anims.size(); i++) {
		if (anims[i] != NULL)
			anims[i]->advanceFrame();
	}
}

void GameSlotPreview::setAnimation(const std::string& name) {
	if (activeAnimation) {
		if (name == activeAnimation->getName())
			return;

		delete activeAnimation;
	}

	activeAnimation = animationSet->getAnimation(name);

	for (unsigned i=0; i < animsets.size(); i++) {
		delete anims[i];
		if (animsets[i])
			anims[i] = animsets[i]->getAnimation(name);
		else
			anims[i] = 0;
	}
}

void GameSlotPreview::addRenders(std::vector<Renderable> &r) {
	if (!stats)
		return;

	for (unsigned i = 0; i < layer_def[stats->direction].size(); ++i) {
		unsigned index = layer_def[stats->direction][i];
		if (anims[index]) {
			Renderable ren = anims[index]->getCurrentFrame(stats->direction);
	unsigned char dir = *direction;

	for (unsigned i = 0; i < layer_def[dir].size(); ++i) {
		unsigned index = layer_def[dir][i];
		if (index < anims.size() && anims[index]) {
			Renderable ren = anims[index]->getCurrentFrame(dir);
			ren.prio = i+1;
			r.push_back(ren);
		}
	}
}

void GameSlotPreview::setStatBlock(StatBlock *_stats) {
	stats = _stats;
	if (stats) {
		direction = &stats->direction;
	}
}

void GameSlotPreview::setPos(Point _pos) {
	pos = _pos;
}

void GameSlotPreview::setDirection(unsigned char dir) {
	static_direction = dir;
	direction = &static_direction;
}

void GameSlotPreview::render() {
	std::vector<Renderable> r;
	addRenders(r);

	for (size_t i = 0; i < r.size(); ++i) {
		if (r[i].image) {
			Rect dest;
			dest.x = pos.x - r[i].offset.x;
			dest.y = pos.y - r[i].offset.y;
			render_device->render(r[i], dest);
		}
	}
}

void GameSlotPreview::loadDefaultGraphics() {
	if (!stats) {
		Utils::logError("GameSlotPreview: StatBlock is not set. Can't load default graphics.");
		return;
	}

	default_gfx.clear();

	// fall back to default if it exists
	for (size_t i = 0; i < layer_reference_order.size(); ++i) {
		bool exists = Filesystem::fileExists(mods->locate("animations/avatar/" + stats->gfx_base + "/default_" + layer_reference_order[i] + ".txt"));
		if (exists) {
			default_gfx.push_back("default_" + layer_reference_order[i]);
		}
		else if (layer_reference_order[i] == "head") {
			default_gfx.push_back(stats->gfx_head);
		}
		else {
			default_gfx.push_back("");
		}
	}

}

void GameSlotPreview::loadGraphicsFromInventory(MenuInventory* menu_inv) {
	std::vector<std::string> preview_gfx = default_gfx;

	if (items && menu_inv) {
		size_t storage_size = static_cast<size_t>(menu_inv->inventory[MenuInventory::EQUIPMENT].getSlotNumber());
		for (size_t i = 0; i < storage_size; ++i) {
			ItemID item_id = menu_inv->inventory[MenuInventory::EQUIPMENT].storage[i].item;

			if (item_id == 0)
				continue;

			if (!items->isValid(item_id) || !items->items[item_id]->has_name) {
				// TODO error msg?
				continue;
			}

			if (!menu_inv->isActive(i)) {
				continue;
			}

			if (!layer_reference_order.empty()) {
				std::vector<std::string>::iterator found = find(layer_reference_order.begin(), layer_reference_order.end(), items->items[item_id]->type);
				if (found != layer_reference_order.end()) {
					size_t preview_index = distance(layer_reference_order.begin(), found);
					if (preview_index < preview_gfx.size())
						preview_gfx[preview_index] = items->items[item_id]->gfx;
				}
			}
		}
	}

	loadGraphics(preview_gfx);
}

GameSlotPreview::~GameSlotPreview() {
	anim->decreaseCount("animations/hero.txt");
	if (activeAnimation)
		delete activeAnimation;

	for (unsigned int i=0; i<animsets.size(); i++) {
		if (animsets[i])
			anim->decreaseCount(animsets[i]->getName());
		delete anims[i];
	}
	anim->cleanUp();
}

**Reply by dorkster** [reply]
The code for the inventory preview is already in the copy of the engine that you cloned. No need to add anything.

You can restore GameSlotPreview.cpp to it's original state by running this from the flare-engine directory:
```
git checkout src/GameSlotPreview.cpp
```

In fact, now is probably a good time to learn the basics of `git`. It's pretty essential if you want to maintain your own fork of the engine code. Here's a few commands to get you started (you should still look up a proper tutorial for git):

```
# show a list of which files have been altered / staged for commit
git status

# show the changes in detail
git diff

# create a new branch with the label of kingdoms_stone
git checkout -b "kingdoms_stone"

# include the file src/Flare.rc in the next commit
git add src/Flare.rc

# commit the changes with a descriptive message
git commit -m "Windows EXE info for Kingdom's Stone"
```

-----

Anyway, once you're back to being able to build the engine, you only need to added the following lines to your mod's `menus/inventory.txt` (you might need to adjust the preview_pos):

```
preview_enabled=true
preview_pos=56,166
```

**Reply by BuzzLEquerre** [reply]
Ok thanks! everything works I managed to get my .exe file with the right name, and the inventory layout works great, only is it possible to enlarge the player animation? It would be a small bonus but if the solution is in the game engine don't bother you've already spent a lot of time for me x)
![Capture d'écran 2024-09-01 152642](https://github.com/user-attachments/assets/69bdc189-6993-4925-a533-7c3ac8426a7e)

**Reply by dorkster** [reply]
The engine would need some changes to support scaling up the animated sprites. So for now, you'll just have to live with it being the same size as the in-world character.

**Reply by BuzzLEquerre** [reply]
Okok, no problem I'll reduce the size of the rectangle, but what you did is fantastic ! Really a huge thank you ! ;)
Besides I was thinking that since you helped me a lot, would you agree that I mention you in the game credits ?
Because people who will play my game need to know that I did not code the game engine.

**Reply by dorkster** [reply]
You should include the full engine credits. This means shipping with the `CREDITS.engine.txt` file that comes with the engine.

Then there's the in-game cutscene (the one that plays if you click "Credits" on the title screen). This is loaded from `cutscenes/credits.txt` within the mods. The easiest way to add your credits to the existing ones would be to have `cutscenes/credits.txt` in your mod look like this:

```
APPEND

[vscroll]
# your credits go here
```

This will combine the engine credits from the `default` mod, the art credits from the `fantasycore` mod (since it looks like your game will depend on it), and your mod (in that order). If you want to change the order, you'll need to copy the credits cutscenes from the `default` and `fantasycore` mods and give them unique names. Then your `cutscenes/credits.txt` file can look like this:

```
[vscroll]
# your credits here

INCLUDE cutscenes/credits_fantasycore.txt
INCLUDE cutscenes/credits_engine.txt
```
